deployment/ccip/changeset/cs_deploy_token_pools.go

```diff
@@ -0,0 +1,497 @@
+	if err := t.RemoteChainsToAdd.Validate(); err != nil {
+		return fmt.Errorf("failed to validate remote chains config: %w", err)
+	}
+	if t.Type != BurnMintTokenPool && t.Type != BurnFromMintTokenPool && t.Type != BurnWithFromMintTokenPool && t.Type != LockReleaseTokenPool {
+		return fmt.Errorf("%s is not a valid token pool type", t.Type)
+	}
+	if t.Type == LockReleaseTokenPool && t.AcceptLiquidity == nil {
+		return errors.New("accept liquidity must be defined for lock release pools")
+	}
+	if t.Type != LockReleaseTokenPool && t.AcceptLiquidity != nil {
+		return errors.New("accept liquidity must be nil for burn mint pools")
+	}
+	return nil
+}
+
+// DeployTokenPoolContractsConfig is the configuration for the DeployTokenPoolContracts changeset.
+type DeployTokenPoolContractsConfig struct {
+	Symbol              TokenSymbol
+	TimelockDelay       time.Duration
+	ExistingPoolUpdates map[uint64]BaseTokenPoolInput
+	NewPools            map[uint64]NewTokenPoolInput
+}
+
+func (c DeployTokenPoolContractsConfig) Validate() error {
+	seenChains := make(map[uint64]struct{})
+	for chainSelector, chainConfig := range c.ExistingPoolUpdates {
+		seenChains[chainSelector] = struct{}{}
+		if err := chainConfig.Validate(); err != nil {
+			return fmt.Errorf("chain with selector %d is invalid: %w", chainSelector, err)
+		}
+	}
+	for chainSelector, chainConfig := range c.NewPools {
+		if _, ok := seenChains[chainSelector]; ok {
+			return errors.New("chain overlap exists between new pools and updates to existing pools")
+		}
+		if err := chainConfig.Validate(); err != nil {
+			return fmt.Errorf("chain with selector %d is invalid: %w", chainSelector, err)
+		}
+	}
+
+	return nil
+}
+
+// DeployTokenPoolContract deploys & configures new pools for a given token across multiple chains.
+// The changeset will first deploy new token pools and transfer ownership of the pools to the Timelock.
+// The outputted MCMS proposal will apply chain updates on each token pool and set new pools on the TokenAdminRegistry.
+func DeployTokenPoolContracts(env deployment.Environment, c DeployTokenPoolContractsConfig) (deployment.ChangesetOutput, error) {
+	if err := c.Validate(); err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("invalid DeployTokenPoolContractsConfig: %w", err)
+	}
+	newAddresses := deployment.NewMemoryAddressBook()
+
+	state, err := LoadOnchainState(env)
+	if err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("failed to load onchain state: %w", err)
+	}
+
+	tokenChainConfigs := make(map[uint64]TokenChainConfig)
+
+	// Deploy new token pools & transfer ownership of each new pool to the timelock
+	for chainSelector, chainConfig := range c.NewPools {
+		chainEnv, ok := env.Chains[chainSelector]
+		if !ok {
+			return deployment.ChangesetOutput{}, fmt.Errorf("no chain with selector %d found in environment", chainSelector)
+		}
+		chainState := state.Chains[chainSelector] // state is derived from env, no need to re-check
```

Comment:
though unlikely, it's safe to have a check here for the existence of chain in state, as you are getting `chainSelector` input from user